### ISSUE TITLE
Rust "could not find root" panic when using virtual workspaces

### ISSUE THREAD

**Issue by ts826848** [astral-sh/uv#6386]
Platform: macOS 12.7.6
Version: uv 0.3.0 (dd1934c9c 2024-08-20)

I'm working on migrating a new project from Rye to uv. This project uses Rye's virtual projects since it only uses Python for helper scripts and not for producing any artifacts, so I tried using uv's virtual workspace as the closest available equivalent.

However, I'm running into Rust panics when attempting to add/remove packages. These panics have been rather inconsistent - sometimes I can add/remove a few packages successfully, but they seem to inevitably crop up. I'm not sure exactly what causes them; while the Rust error message says "could not find root", the fact that prior commands (sometimes) succeeded show that the root should be findable and it's unclear to me exactly why the project root is no longer findable, especially since the debug information says it's found.

One example session:

```
pt@rp uv-test % ls -a
./  ../
pt@rp uv-test % uv init --virtual --no-readme
Initialized workspace `uv-test`
pt@rp uv-test % uv lock --verbose
DEBUG uv 0.3.0
DEBUG Found workspace root: `/private/tmp/uv-test`
DEBUG Searching for Python interpreter in managed installations or system path
DEBUG Searching for managed installations at `/Users/pt/.local/share/uv/python`
DEBUG Found `cpython-3.12.5-macos-x86_64-none` at `/usr/local/bin/python3` (search path)
Using Python 3.12.5 interpreter at: /usr/local/opt/python@3.12/bin/python3.12
DEBUG No `requires-python` in virtual-only workspace. Defaulting to `>=3.12`.
DEBUG Using request timeout of 30s
DEBUG Starting clean resolution
DEBUG Solving with installed Python version: 3.12.5
DEBUG Solving with target Python version: >=3.12
DEBUG Tried 0 versions:
DEBUG Split universal resolution took 0.002s
Resolved 0 packages in 14ms
pt@rp uv-test % uv add --dev ipython --verbose
DEBUG uv 0.3.0
DEBUG Found project root: `/private/tmp/uv-test`
DEBUG Searching for Python interpreter in managed installations or system path
DEBUG Searching for managed installations at `/Users/pt/.local/share/uv/python`
DEBUG Found `cpython-3.12.5-macos-x86_64-none` at `/usr/local/bin/python3` (search path)
Using Python 3.12.5 interpreter at: /usr/local/opt/python@3.12/bin/python3.12
Creating virtualenv at: .venv
DEBUG Using request timeout of 30s
DEBUG No `requires-python` in virtual-only workspace. Defaulting to `>=3.12`.
DEBUG Using request timeout of 30s
DEBUG Existing `uv.lock` satisfies workspace requirements
Resolved 0 packages in 0.23ms
thread 'main' panicked at crates/uv-resolver/src/lock.rs:455:22:
could not find root
stack backtrace:
   0:        0x105017a23 - __mh_execute_header
   1:        0x104c696cb - __mh_execute_header
   2:        0x104fe108e - __mh_execute_header
   3:        0x10501bb01 - __mh_execute_header
   4:        0x10501b379 - __mh_execute_header
   5:        0x10501d152 - __mh_execute_header
   6:        0x10501be42 - __mh_execute_header
   7:        0x10501bd99 - __mh_execute_header
   8:        0x10501bd86 - __mh_execute_header
   9:        0x105d1bc52 - __mh_execute_header
  10:        0x105d1c27a - __mh_execute_header
  11:        0x105a45537 - __mh_execute_header
  12:        0x1051678cd - __mh_execute_header
  13:        0x105172af0 - __mh_execute_header
  14:        0x105143116 - __mh_execute_header
  15:        0x1051f3f77 - __mh_execute_header
  16:        0x1051f33a9 - __mh_execute_header
  17:        0x1053b82b9 - __mh_execute_header
  18:        0x1056f3c8b - __mh_execute_header
  19:        0x1052611a6 - __mh_execute_header
  20:        0x1056f3697 - __mh_execute_header
```

**Comment by charliermarsh** [comment]
I believe this is fixed in v0.3.1 which is building now. Do you mind trying again in a bit?

**Comment by charliermarsh** [comment]
Let me know if you still hit this with v0.3.1.

**Comment by lynnporu** [comment]
I had this error when I renamed the project in pyproject.toml manually. Project name in uv.lock remained the same and this made uv fail. The solution was to remove uv.lock and run `uv sync` again.

**Comment by zanieb** [comment]
@lynnporu I cannot reproduce that. Are you on the latest version? Can you share a minimal reproduction?

**Comment by lucinvitae** [comment]
https://github.com/astral-sh/uv/issues/6386#issuecomment-2607986724 <-- I was able to reproduce a similar error after upgrading from uv 0.4.6 to uv 0.5.25 when changing `requires-python` (from `"==3.11.*"` to `"==3.12.*"`) for my project. I didn't need to rename the project to repro the error, just those two changes.

After making those changes, no matter what I tried, the `error: Could not find root package` error appeared:
```
DEBUG uv 0.5.25 (Homebrew 2025-01-28)
DEBUG Found project root: `/Users/me/w/repo/apps/my-app`
DEBUG Project is contained in non-workspace project: `/Users/me/w/repo`
DEBUG No workspace root found, using project root
DEBUG Reading Python requests from version file at `/Users/me/w/repo/apps/my-app/.python-version`
DEBUG Using Python request `3.12.8` from version file at `.python-version`
DEBUG The virtual environment's Python version satisfies `3.12.8`
DEBUG Using request timeout of 30s
DEBUG Ignoring existing lockfile due to mismatched source: `my-app` (expected: `virtual`)
DEBUG Found static `pyproject.toml` for: my-app @ file:///Users/me/w/repo/apps/my-app
DEBUG Project is contained in non-workspace project: `/Users/me/w/repo`
DEBUG No workspace root found, using project root
DEBUG Found static `pyproject.toml` for: auth @ file:///Users/me/w/repo/library/cies_auth
DEBUG Found static `pyproject.toml` for: schema @ file:///Users/me/w/repo/library/api_schema_lib
DEBUG Project is contained in non-workspace project: `/Users/me/w/repo`
DEBUG No workspace root found, using project root
DEBUG Project is contained in non-workspace project: `/Users/me/w/repo`
DEBUG No workspace root found, using project root
DEBUG Project is contained in non-workspace project: `/Users/me/w/repo`
DEBUG No workspace root found, using project root
DEBUG Project is contained in non-workspace project: `/Users/me/w/repo`
DEBUG No workspace root found, using project root
DEBUG Solving with installed Python version: 3.12.8
DEBUG Solving with target Python version: ==3.12.*
DEBUG Solving split (false) (requires-python: RequiresPython { specifiers: VersionSpecifiers([VersionSpecifier { operator: EqualStar, version: "3.12" }]), range: RequiresPythonRange(LowerBound(Included("3.12")), UpperBound(Excluded("3.13"))) })
DEBUG Tried 0 versions:
DEBUG split `false` resolution took 0.000s
DEBUG Distinct solution for split (false) with 0 packages
Resolved in 2ms
error: Could not find root package `my-app`
```

Solution as above: delete `uv.lock` and regenerate with `uv lock`. 

Now all is well for the project again. I don't have a minimal reproducible example from start to finish, but I will try to get one, the above comment helped me. Unfortunately, it also means all 3rdparty deps get updated for the project, which isn't necessarily desirable.

P.S. uv is an awesome tool! 👏
